api/proto/teleport/legacy/types/types.proto

```diff
@@ -4743,3 +4743,106 @@ enum RequireMFAType {
+
+  // Metadata is the resource metadata.
+  Metadata Metadata = 4 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "metadata"
+  ];
+
+  PluginSpecV1 Spec = 5 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "spec"
+  ];
+
+  PluginStatusV1 Status = 6 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "status"
+  ];
+
+  // Credentials are "live" credentials to the 3rd party API.
+  // These are considered secrets.
+  PluginCredentialsV1 Credentials = 7 [(gogoproto.jsontag) = "credentials,omitempty"];
+}
+
+message PluginSpecV1 {
+  option (gogoproto.equal) = true;
+
+  // Settings contain provider-specific plugin options.
+  oneof Settings {
+    // Settings for the Slack access plugin
+    PluginSlackAccessSettings SlackAccessPlugin = 1 [(gogoproto.jsontag) = "slack_access,omitempty"];
+  }
+}
+
+message PluginSlackAccessSettings {
+  option (gogoproto.equal) = true;
+
+  string FallbackChannel = 1 [(gogoproto.jsontag) = "fallback_channel"];
+}
+
+message PluginInitialCredentialsV1 {
+  oneof Credentials {
+    PluginOAuth2AuthorizationCodeCredentials OAuth2AuthorizationCode = 1 [(gogoproto.jsontag) = "oauth2_authorization_code,omitempty"];
+  }
+}
+
+message PluginOAuth2AuthorizationCodeCredentials {
+  string AuthorizationCode = 1 [(gogoproto.jsontag) = "authorization_code"];
+  string RedirectURI = 2 [(gogoproto.jsontag) = "redirect_uri"];
+}
+
+// PluginStatus is the user-facing status for the plugin instance.
+message PluginStatusV1 {
+  PluginStatusCode Code = 1 [(gogoproto.jsontag) = "code"];
+}
+
+enum PluginStatusCode {
+  // UNKNOWN is the default value when the plugin has not reported its status yet.
+  UNKNOWN = 0;
+  // RUNNING means the plugin reports running successfully.
+  RUNNING = 1;
+  // OTHER_ERROR indicates that an otherwise-unspecified error has been encountered.
+  OTHER_ERROR = 2;
+  // UNAUTHORIZED indicates that plugin is not able to authenticate to the 3rd party API.
+  // This could be a result of e.g. the user revoking the authorization on the API provider's side.
+  UNAUTHORIZED = 3;
+
+  // SLACK_NOT_IN_CHANNEL is a Slack-specific status code that indicates
+  // that the bot has not been invited to a channel that it is configured to post in.
+  SLACK_NOT_IN_CHANNEL = 10;
+  reserved 11 to 19; // future Slack-specific codes
+}
+
+// PluginCredentialsV1 represents "live" credentials
+// that are used by the plugin to authenticate to the 3rd party API.
+message PluginCredentialsV1 {
+  oneof Credentials {
+    PluginOAuth2AccessTokenCredentials OAuth2AccessToken = 1 [(gogoproto.jsontag) = "oauth2_access_token,omitempty"];
+  }
+}
+
+message PluginOAuth2AccessTokenCredentials {
+  string AccessToken = 1;
+  string RefreshToken = 2;
+  google.protobuf.Timestamp ExpiresAt = 3 [
+    (gogoproto.stdtime) = true,
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "expires_at,omitempty"
+  ];
+}
+
+// PluginList represents a list of plugin resources
```

Comment:
Do we ever expect to have more than 50 or so plugin configurations? If so, we shouldn't have a "Get all plugins" endpoint, but rather a "Stream all plugins" one.

Reply:
I would not expect an average cluster to have many individual plugin instances. I imagine a customer would almost always have one instance per service (e.g. 1 Slack integration, 1 PagerDuty integration and 1 email integration). Perhaps a few integrations per service (e.g. connecting several Slack workspaces), but not many dozens :slightly_smiling_face: In the near future, [routing rules](https://github.com/gravitational/teleport/blob/master/rfd/0087-access-request-notification-routing.md) are planned, which would allow some flexibility in notification routing in a single plugin instance.

Reply:
It's fine like this then 👍